presto-plugin-toolkit/src/main/java/com/facebook/presto/security/RefreshingAccessControlUtils.java

```diff
@@ -0,0 +1,102 @@
+{
+    private RefreshingAccessControlUtils()
+    {
+    }
+
+    public static SystemAccessControl refreshingSystemAccessControl(Supplier<SystemAccessControl> factory, Duration refreshPeriod)
+    {
+        return new DelegateSystemAccessControl(new Refresher<>(factory, () -> true, refreshPeriod));
+    }
+
+    public static SystemAccessControl refreshingSystemAccessControl(Supplier<SystemAccessControl> factory, Supplier<Boolean> refreshCondition, Duration refreshPeriod)
+    {
+        return new DelegateSystemAccessControl(new Refresher<>(factory, refreshCondition, refreshPeriod));
+    }
+
+    public static ConnectorAccessControl refreshingConnectorAccessControl(Supplier<ConnectorAccessControl> factory, Duration refreshPeriod)
+    {
+        return new DelegateConnectorAccessControl(new Refresher<>(factory, () -> true, refreshPeriod));
+    }
+
+    public static ConnectorAccessControl refreshingConnectorAccessControl(Supplier<ConnectorAccessControl> factory, Supplier<Boolean> refreshCondition, Duration refreshPeriod)
+    {
+        return new DelegateConnectorAccessControl(new Refresher<>(factory, refreshCondition, refreshPeriod));
+    }
+
+    public static Supplier<Boolean> wasFileModified(Path path)
+    {
+        AtomicLong configFileNameLastModification = new AtomicLong();
+        return () -> {
+            long lastModification = path.toFile().lastModified();
+            return configFileNameLastModification.get() != lastModification &&
+                    configFileNameLastModification.compareAndSet(configFileNameLastModification.get(), lastModification);
+        };
+    }
+
+    private static final class Refresher<T>
+            implements Supplier<T>
+    {
+        private final Supplier<T> supplier;
+        private final Supplier<Boolean> refreshCondition;
+        private final Duration refreshPeriod;
+        private final AtomicReference<T> object = new AtomicReference<>();
+        private final AtomicLong lastRefreshNanos = new AtomicLong(System.nanoTime());
+
+        public Refresher(Supplier<T> supplier, Supplier<Boolean> refreshCondition, Duration refreshPeriod)
+        {
+            this.supplier = requireNonNull(supplier, "supplier is null");
+            this.refreshCondition = requireNonNull(refreshCondition, "refreshCondition is null");
+            this.refreshPeriod = requireNonNull(refreshPeriod, "refreshPeriod is null");
+            object.set(this.supplier.get());
+        }
+
+        @Override
+        public T get()
+        {
+            if (Duration.nanosSince(lastRefreshNanos.get()).compareTo(refreshPeriod) > 0) {
+                refresh();
+            }
+            return object.get();
+        }
+
+        private void refresh()
+        {
+            if (!lastRefreshNanos.compareAndSet(lastRefreshNanos.get(), System.nanoTime())) {
+                return;
+            }
+            if (refreshCondition.get()) {
+                object.set(supplier.get());
```

Comment:
You may end up calling `supplier` by 2 threads concurrently. Intentional?

Reply:
this line is actually almost "raceful equivalent" of `lastRefreshNanos.set(System.nanoTime()); continue;`, so any number of threads can pass this line